File: clients/python/smg_client/api/workers.py
"""Workers API."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from smg_client.types import WorkerApiResponse, WorkerInfo, WorkerListResponse

if TYPE_CHECKING:
    from smg_client._transport import AsyncTransport, SyncTransport


class SyncWorkers:
    """Synchronous workers API."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def create(self, **kwargs: Any) -> WorkerApiResponse:
        resp = self._transport.request("POST", "/workers", json=kwargs)
        return WorkerApiResponse.model_validate_json(resp.content)

    def list(self) -> WorkerListResponse:
        resp = self._transport.request("GET", "/workers")
        return WorkerListResponse.model_validate_json(resp.content)

    def get(self, worker_id: str) -> WorkerInfo:
        resp = self._transport.request("GET", f"/workers/{worker_id}")
        return WorkerInfo.model_validate_json(resp.content)

    def update(self, worker_id: str, **kwargs: Any) -> WorkerApiResponse:
        resp = self._transport.request("PUT", f"/workers/{worker_id}", json=kwargs)
        return WorkerApiResponse.model_validate_json(resp.content)

    def delete(self, worker_id: str) -> WorkerApiResponse:
        resp = self._transport.request("DELETE", f"/workers/{worker_id}")
        return WorkerApiResponse.model_validate_json(resp.content)


class AsyncWorkers:
    """Async workers API."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def create(self, **kwargs: Any) -> WorkerApiResponse:
        resp = await self._transport.request("POST", "/workers", json=kwargs)
        return WorkerApiResponse.model_validate_json(resp.content)

    async def list(self) -> WorkerListResponse:
        resp = await self._transport.request("GET", "/workers")
        return WorkerListResponse.model_validate_json(resp.content)

    async def get(self, worker_id: str) -> WorkerInfo:
        resp = await self._transport.request("GET", f"/workers/{worker_id}")
        return WorkerInfo.model_validate_json(resp.content)

    async def update(self, worker_id: str, **kwargs: Any) -> WorkerApiResponse:
        resp = await self._transport.request(
            "PUT", f"/workers/{worker_id}", json=kwargs
        )
        return WorkerApiResponse.model_validate_json(resp.content)

    async def delete(self, worker_id: str) -> WorkerApiResponse:
        resp = await self._transport.request("DELETE", f"/workers/{worker_id}")
        return WorkerApiResponse.model_validate_json(resp.content)

File: clients/rust/src/api/workers.rs
use openai_protocol::worker::{
    WorkerApiResponse, WorkerInfo, WorkerListResponse, WorkerSpec, WorkerUpdateRequest,
};

use crate::{transport::Transport, SmgError};

/// Workers API (`/workers`).
pub struct Workers {
    transport: Transport,
}

impl Workers {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    /// Register a new worker.
    pub async fn create(&self, spec: &WorkerSpec) -> Result<WorkerApiResponse, SmgError> {
        let resp = self.transport.post("/workers", spec).await?;
        let body = resp.text().await.map_err(SmgError::Connection)?;
        serde_json::from_str(&body).map_err(SmgError::from)
    }

    /// List all registered workers.
    pub async fn list(&self) -> Result<WorkerListResponse, SmgError> {
        let resp = self.transport.get("/workers").await?;
        let body = resp.text().await.map_err(SmgError::Connection)?;
        serde_json::from_str(&body).map_err(SmgError::from)
    }

    /// Get details for a specific worker.
    pub async fn get(&self, worker_id: &str) -> Result<WorkerInfo, SmgError> {
        let resp = self.transport.get(&format!("/workers/{worker_id}")).await?;
        let body = resp.text().await.map_err(SmgError::Connection)?;
        serde_json::from_str(&body).map_err(SmgError::from)
    }

    /// Update a worker's configuration.
    pub async fn update(
        &self,
        worker_id: &str,
        request: &WorkerUpdateRequest,
    ) -> Result<WorkerApiResponse, SmgError> {
        let resp = self
            .transport
            .put(&format!("/workers/{worker_id}"), request)
            .await?;
        let body = resp.text().await.map_err(SmgError::Connection)?;
        serde_json::from_str(&body).map_err(SmgError::from)
    }

    /// Remove a worker.
    pub async fn delete(&self, worker_id: &str) -> Result<WorkerApiResponse, SmgError> {
        let resp = self
            .transport
            .delete(&format!("/workers/{worker_id}"))
            .await?;
        let body = resp.text().await.map_err(SmgError::Connection)?;
        serde_json::from_str(&body).map_err(SmgError::from)
    }
}
